kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs

```diff
@@ -1943,6 +1944,47 @@ impl<'tcx> GotocCtx<'tcx> {
         let zero = Type::size_t().zero();
         cast_ptr.rem(align).eq(zero)
     }
+
+    /// Swaps the memory contents pointed to by arguments `x` and `y`, respectively, which is
+    /// required for the `typed_swap` intrinsic.
+    ///
+    /// The standard library API requires that `x` and `y` are readable and writable as their
+    /// (common) type (which auto-generated checks for dereferencing will take care of), and the
+    /// memory regions pointed to must be non-overlapping.
+    pub fn codegen_swap(&mut self, fargs: Vec<Expr>, loc: Location) -> Stmt {
+        // if(same_object(x, y)) {
+        //   assert(x + 1 <= y || y + 1 <= x);
+        //   assume(x + 1 <= y || y + 1 <= x);
```

Comment:
Is this how the check is implemented in CBMC? I don't know if we want to re-implement the check because there are a bunch of corner cases we could be missing. For example, will this properly handle ZSTs? What about aliasing?

Because of that, I'm wondering if it's better to replace the assignments by calls to `memcpy` which already include those checks IIRC.

Reply:
1) Yes, CBMC's `memcpy` uses the same checks.
2) Aliasing is handled via the `same_object` condition.
3) Using `memcpy` would introduce overhead as we have type information here and don't need to go for byte-by-byte copying. No need to compute type sizes (in bytes), which in turn may raise questions about possible arithmetic overflow.